app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/MainViewModel.kt

```diff
@@ -0,0 +1,157 @@
+package nerd.tuxmobil.fahrplan.congress.schedule
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import info.metadude.android.eventfahrplan.commons.livedata.SingleLiveEvent
+import info.metadude.android.eventfahrplan.commons.logging.Logging
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import nerd.tuxmobil.fahrplan.congress.changes.ChangeStatistic
+import nerd.tuxmobil.fahrplan.congress.models.Meta
+import nerd.tuxmobil.fahrplan.congress.net.ParseResult
+import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository
+import nerd.tuxmobil.fahrplan.congress.repositories.ExecutionContext
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.FetchFailure
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.FetchSuccess
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.Fetching
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.InitialFetching
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.InitialParsing
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.ParseFailure
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.ParseSuccess
+import nerd.tuxmobil.fahrplan.congress.repositories.LoadScheduleState.Parsing
+import nerd.tuxmobil.fahrplan.congress.schedule.observables.LoadScheduleUiState
+import nerd.tuxmobil.fahrplan.congress.schedule.observables.ScheduleChangesParameter
+
+class MainViewModel(
+
+    private val repository: AppRepository,
+    private val executionContext: ExecutionContext,
+    private val logging: Logging
+
+) : ViewModel() {
+
+    val loadScheduleUiState = SingleLiveEvent<LoadScheduleUiState>()
+    val scheduleChangesParameter = SingleLiveEvent<ScheduleChangesParameter>()
+    val showAbout = SingleLiveEvent<Meta>()
+    val openSessionDetails = SingleLiveEvent<Unit>()
+
+    private val mutableFetchFailure = MutableLiveData<FetchFailure?>()
+    val fetchFailure: LiveData<FetchFailure?> = mutableFetchFailure
```

Comment:
@cketti I think I misunderstood you here. This has to be a `SingleLiveEvent` to prevent that the dialog or toast reappears after an orientation change. Right?

Reply:
If the choice is between `LiveData` and `SingleLiveEvent`, then yes, this should probably be a `SingleLiveEvent`.

Things should get easier if you strictly distinguish between UI state and UI action events. You'd use a UI action event to display a toast, but update the UI state to change the view hierarchy to display e.g. an error message. If you include "state" and "action" or "event" in names it should be easier to see which is what.

`SingleLiveEvent` is not a great fit for events, because there's no queue. But it'll do if you use one `SingleLiveEvent` per event type and skipping intermediate events while the UI isn't observing is fine.

Reply:
I see. Good wording definitely helps. Currently, both `AlertDialog` and `Toast` launched by the `ErrorMessage.Factory`. I keep the wording for later when one works on the overall error messaging mechanism and UI appearance.

Also, it will important to clearly define what "event" and "action" are associated with - since it is not common sense, afaik.

Reply:
:heavy_check_mark: Fixed in ee444b44fe48f47759782936c12ef7532bd4457a.